### ISSUE TITLE
Suggestion: strict mode should detect yield-returning inside a lock

### ISSUE THREAD

**Issue by WizardBrony** [dotnet/roslyn#46004]
For example, the following code fails:

```cs
public class C
{
    private readonly object _lock = new object();

    public IEnumerable<int> Iterator()
    {
        lock (_lock)
            yield return 0;
    }
}

var c = new C();
foreach (var i in c.Iterator())
    await Task.Yield();
```

The compiler should generate an error (or at least a warning) informing the user that the currently-executing thread after yield-returning may not be the same thread that the lock was taken on. Considering that the compiler already prevents awaiting in the body of a lock statement (Error CS1996) for the same reason, I feel like this addition would follow suit.

**Comment by jcouv** [comment]
Tagging @jjones. I think we can close this now as LDM discussed this and concluded that `yield return` should be allowed in a `lock` as it is in a `try/finally`.
See https://github.com/dotnet/roslyn/pull/74024 for context
